### PR TITLE
AIILG-442 Backend tests users

### PR DESCRIPTION
**Pull request by imad-chowdhury-dluhc**
# Overview
Adds tests for `backend/api/routes/users.py` and `backend/api/dependencies/get_current_user.py`

## JIRA Ticket
https://mhclgdigital.atlassian.net/browse/AIILG-442

## Changes
- 2 test files

## Acceptance Criteria (AC)
- [ ] AC1: test `users` API route
- [ ] AC2: test `get_current_user` dependency

---

## Testing

**Manual**:
- `poetry run pytest tests/backend/api/routes/test_users.py --no-cov` for `users.py`
- `poetry run pytest tests/backend/api/dependencies/get_current_user.py --no-cov `  for `get_current_users.py`

## Notes / Additional Information (optional)
- `get_sessions.py` dependency is not tested - does this need to be? It's a very small wrapper.
- Currently `test_users.py` requires a queue service to be running. Is this acceptable or does this need to be mocked?

### CODE REVIEW

File: tests/backend/api/dependencies/get_current_user.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    session.commit = AsyncMock()
+    session.refresh = AsyncMock()
+    session.exec = AsyncMock()
+
+    return session
+
+
+@pytest.mark.asyncio
+async def test_get_current_user_existing_user(monkeypatch, session):
+    mock_user = Mock(email="test@example.com")
```
Comment (Patryk-Kuchta) [inline tests/backend/api/dependencies/get_current_user.py]: I think in places like this we should be using the real class with mocked values rather than mocking the whole thing, just import and use this:

```
from common.database.postgres_models import User
```

File: tests/backend/api/dependencies/get_current_user.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    )
+
+    user = await get_current_user(
+        session=session,
+        x_amzn_oidc_data="token",
+    )
+
+    # check emails match and no new user is created
+    assert user == mock_user
+    session.commit.assert_not_called()
```
Comment (Patryk-Kuchta) [inline tests/backend/api/dependencies/get_current_user.py]: Can we assert that the correct token was used in the call too?

File: tests/backend/api/dependencies/get_current_user.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        lambda _: mock_auth_info,  # token unused in test
+    )
+
+    user = await get_current_user(
+        session=session,
+        x_amzn_oidc_data="token",
+    )
+
+    # check emails match and no new user is created
+    assert user == mock_user
```
Comment (Patryk-Kuchta) [inline tests/backend/api/dependencies/get_current_user.py]:
```suggestion
    assert user.email == "test@example.com"
    assert user is mock_user
```
This should be a bit more water-tight

File: tests/backend/api/routes/conftest.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+@pytest.fixture
+def mock_user():
+    return User(
+        id=uuid4(),
+        email="test@local-transcribe.com",
+        data_retention_days=30,
+        created_datetime=datetime.now(UTC),
+        updated_datetime=datetime.now(UTC),
+    )
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/conftest.py:21]: Nice :D

Comment (Patryk-Kuchta) [reply tests/backend/api/routes/conftest.py:21]: This is exactly the pattern I was thinking of when writing this comment https://github.com/communitiesuk/localai-local-transcribe/pull/125/changes#r3050385649

File: tests/backend/api/routes/test_users.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    override_session,
+):
+    async with get_test_client() as ac:
+        response = await ac.patch(
+            "/users/data-retention",
+            json={"data_retention_days": 0},
+        )
+
+    assert response.status_code == 400
+    assert "at least 1 day" in response.json()["detail"]
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_users.py]: Would it be possible to match the full error string here?

File: tests/backend/api/routes/test_users.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+@pytest.mark.asyncio(loop_scope="session")
+async def test_update_data_retention_invalid(
+    override_user,
+    override_session,
+):
+    async with get_test_client() as ac:
+        response = await ac.patch(
+            "/users/data-retention",
+            json={"data_retention_days": 0},
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_users.py:65]: aren't we missing a test for None (i.e. Indefinite?)

File: tests/backend/api/routes/test_users.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            "/users/data-retention",
+            json={"data_retention_days": 10},
+        )
+
+    assert response.status_code == 200
+    data = response.json()
+
+    # check retention data updated and returns updated user
+    assert mock_user.data_retention_days == 10
+    assert data["data_retention_days"] == 10
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_users.py]: I think this is a stricter assert (prevents success if 10 ever becomes our default)
```suggestion
    original_retention = mock_user.data_retention_days
    target_retention = 10
    
    async with get_test_client() as ac:
        response = await ac.patch(
            "/users/data-retention",
            json={"data_retention_days": target_retention},
        )
 
    assert response.status_code == 200
    data = response.json()
 
    assert original_retention == 30  # Confirm it started at 30
    assert mock_user.data_retention_days == target_retention 
    assert data["data_retention_days"] == target_retention
```

File: tests/backend/api/routes/test_users.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    # check db has been used
+    mock_session.commit.assert_awaited_once()
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_users.py:53]: could we also assert for the modification timestamps being updated when retention changes

File: tests/backend/api/routes/test_users.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    override_user,
+    override_session,
+    mock_user,
+    mock_session,
+):
+    async with get_test_client() as ac:
+        response = await ac.patch(
+            "/users/data-retention",
+            json={"data_retention_days": 10},
+        )
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_users.py:42]: I really like this test, but I wonder if it modifies the mock user with persistent effect (i.e. all following test will use 10 for retention rather than 30), which would be undesirable. Can you check if that's the case?

File: tests/backend/api/routes/test_users.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+import pytest
+
+from tests.utils import get_test_client
+
+
+def convert_to_datetime(json_datetime: str) -> datetime:
+    return datetime.fromisoformat(json_datetime.replace("Z", "+00:00"))
+
+
+@pytest.mark.asyncio(loop_scope="session")
```
Comment (Patryk-Kuchta) [inline tests/backend/api/routes/test_users.py]: info: What's the rationale being `loop_scope="session"`

**Review by Patryk-Kuchta — CHANGES_REQUESTED** [review]
Good stuff, a few ways of improving the tests in the comments and a few questions :)

File: tests/backend/api/dependencies/get_current_user.py (hunk 1 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/dependencies/get_current_user.py]: [Fixed](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/dependencies/get_current_user.py#L34-L40)

File: tests/backend/api/dependencies/get_current_user.py (hunk 2 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/dependencies/get_current_user.py]: Fixed [here](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/dependencies/get_current_user.py#L59), [here](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/dependencies/get_current_user.py#L82), and [here](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/dependencies/get_current_user.py#L103)

File: tests/backend/api/dependencies/get_current_user.py (hunk 3 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/dependencies/get_current_user.py]: Fixed

File: tests/backend/api/routes/test_users.py (hunk 5 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/routes/test_users.py]: [Fixed](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/routes/test_users.py#L69-L70)

File: tests/backend/api/routes/test_users.py (hunk 6 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/routes/test_users.py:65]: [Fixed](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/routes/test_users.py#L30)

File: tests/backend/api/routes/test_users.py (hunk 7 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/routes/test_users.py]: Fixed [here](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/routes/test_users.py#L35) before we update it so we don't need a variable to keep track, let me know if that doesn't work.

File: tests/backend/api/routes/test_users.py (hunk 8 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/routes/test_users.py:53]: I've added an assert [here](https://github.com/communitiesuk/localai-local-transcribe/blob/7d19a0c07a43a9b9e8bcf46950a6524ef0910aa2/tests/backend/api/routes/test_users.py#L50) to check that the new `updated_datetime` is not equal to the old. Is there a more careful way to check the new `updated_datetime` is what we expect it to be?

File: tests/backend/api/routes/test_users.py (hunk 10 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/routes/test_users.py]: I think this is related to this [question](https://github.com/communitiesuk/localai-local-transcribe/pull/125#discussion_r3050755352) above. I thought `loop_scope="session"` meant that changes in state from one test persist for the rest of the tests in the same session. But when printing out `mock_user`, it seems like the `retention_period` reset to 30 anyways even in other tests. 

However, good catch, I don't think `loop_scope="session"` is doing anything here anyways, we're not connecting to other stuff anyways so each test can be started new without causing any problems. I've removed it.

File: tests/backend/api/routes/test_users.py (hunk 9 above)
Comment (imad-chowdhury-dluhc) [reply tests/backend/api/routes/test_users.py:42]: I've printed out the `mock_user.data_retention_days` across tests and it seems not to persist (resets to 30). 

Do you know if tests happen in the order that they appear in, or does pytest do anything fancy and run them all at once? In which case, my method of just printing out mock_user may not be sufficient is the `test_update_data_retention_success` is the last one to run.

Comment (Patryk-Kuchta) [reply tests/backend/api/routes/test_users.py:42]: Cool, thanks! I am glad you tested that. I check a bit more at it makes sense it isn't persistent :)

**Review by Patryk-Kuchta — APPROVED** [review]
Great work!!!
